lib/errors.js

```diff
@@ -0,0 +1,267 @@
+'use strict'
+
+// Handle creating error event objects to be sent to APM server.
+// https://github.com/elastic/apm-server/blob/master/docs/spec/v2/error.json
+
+const crypto = require('crypto')
+var path = require('path')
+const util = require('util')
+
+const { gatherStackTrace } = require('./stacktraces')
+
+const MYSQL_ERROR_MSG_RE = /(ER_[A-Z_]+): /
+
+// ---- internal support functions
+
+// Default `culprit` to the top of the stack or the highest non `library_frame`
+// frame if such exists
+function culpritFromStacktrace (frames) {
+  if (frames.length === 0) return
+
+  var filename = frames[0].filename
+  var fnName = frames[0].function
+  for (var n = 0; n < frames.length; n++) {
+    if (!frames[n].library_frame) {
+      filename = frames[n].filename
+      fnName = frames[n].function
+      break
+    }
+  }
+
+  return filename ? fnName + ' (' + filename + ')' : fnName
+}
+
+// Infer the node.js module name from the top frame filename, if possible.
+// Here `frames` is a data structure as returned by `parseStackTrace`.
+//
+// Examples:
+//    node_modules/mymodule/index.js
+//                 ^^^^^^^^
+//    node_modules/@myorg/mymodule/index.js
+//                 ^^^^^^^^^^^^^^^
+// or on Windows:
+//    node_modules\@myorg\mymodule\lib\subpath\index.js
+//                 ^^^^^^^^^^^^^^^
+let SEP = path.sep
+if (SEP === '\\') {
+  SEP = '\\' + SEP // Escape this for use in a regex.
+}
+const MODULE_NAME_REGEX = new RegExp(`node_modules${SEP}([^${SEP}]*)(${SEP}([^${SEP}]*))?`)
+function _moduleNameFromFrames (frames) {
+  if (frames.length === 0) return
+  var frame = frames[0]
+  if (!frame.library_frame) return
+  var match = frame.filename.match(MODULE_NAME_REGEX)
+  if (!match) return
+  var moduleName = match[1]
+  if (moduleName && moduleName[0] === '@' && match[3]) {
+    // Normalize the module name separator to '/', even on Windows.
+    moduleName += '/' + match[3]
+  }
+  return moduleName
+}
+
+// Gather properties from `err` to be used for `error.exception.attributes`.
+// If there are no properties to include, then it returns undefined.
+function attributesFromErr (err) {
```

Comment:
REVIEW-NOTE: This is based off stackman.properties() that was used before (identical except for the handling of `key === 'code'` moved here from elsewhere, and formatting).